test/app/App/Panel/AddToken/index.test.js

```diff
@@ -7,162 +7,179 @@ import store from '../../../../../main/store'
 import link from '../../../../../resources/link'
 import AddTokenComponent from '../../../../../dash/App/Tokens/AddToken'
 
-const state = {
-  main: {
-    networks: { 
-      ethereum: {
-        1: {
-          name: 'Mainnet',
-          on: true
-        },
-        137: {
-          name: 'Polygon',
-          on: true
-        }
-      }
-    }
-  }
-}
-
-const chainMetadata = {
-  1: { name: 'Frame Test', symbol: 'FRT', decimals: '18' },
-  137: { name: 'Frame Test on Polygon', symbol: 'mFRT', decimals: '18' }
-}
-
-jest.mock('../../../../../main/store/state', () => () => state)
 jest.mock('../../../../../main/store/persist')
-jest.mock('../../../../../resources/link')
+jest.mock('../../../../../resources/link', () => ({
+  invoke: jest.fn().mockResolvedValue({}),
+  send: jest.fn()
+}))
 
 const AddToken = Restore.connect(AddTokenComponent, store)
 
-const setup = (jsx, options = {}) => ({ user: userEvent.setup(options), ...render(jsx) })
+const setup = (jsx, options = {}) => ({
+  user: userEvent.setup({
+    ...options,
+    advanceTimers: options.advanceTimers || advanceTimers
+  }),
+  ...render(jsx)
+})
 
 const advanceTimers = (ms = 0) => {
   jest.advanceTimersByTime(ms)
   return Promise.resolve()
 }
 
-let rerenderComponent
-
 beforeAll(() => {
   jest.useFakeTimers()
-})
-
-afterAll(() => {
-  jest.useRealTimers()
-})
 
-beforeEach(() => {
-  link.invoke.mockImplementation((channel, contractAddress, chainId) => {
-    if (contractAddress === '0x3432b6a60d23ca0dfca7761b7ab56459d9c964d0') {
-      return Promise.resolve(chainMetadata[chainId])
+  store.addNetwork({
+    id: 1,
+    type: 'ethereum',
+    name: 'Mainnet',
+    explorer: 'https://etherscan.io',
+    symbol: 'ETH',
+    on: true,
+    connection: {
+      primary: { connected: true }
     }
-
-    return Promise.resolve({ name: '', symbol: '', decimals: 0 })
   })
 
-  link.send.mockImplementation((channel, msg, { view, data: { notify, notifyData } }) => {
-    expect(channel).toBe('tray:action')
-    expect(msg).toBe('navDash')
-    expect(view).toBe('tokens')
-    expect(notify).toBe('addToken')
-
-    // simulate nav updating notify data which will pass new props to this component
-    rerenderComponent(
-      <AddToken 
-        activeChains={[
-          { id: 1, name: 'Mainnet', connection: { primary: { connected: true } } },
-          { id: 137, name: 'Polygon', connection: { primary: { connected: true } } }
-        ]} 
-        data={{ notifyData }}
-      />
-    )
+  store.setPrimary('ethereum', 1, { connected: false })
+  store.activateNetwork('ethereum', 1, true)
+
+  store.addNetwork({
+    id: 137,
+    type: 'ethereum',
+    name: 'Polygon',
+    explorer: 'https://polygonscan.com',
+    symbol: 'MATIC',
+    on: true,
+    connection: {
+      primary: { connected: true }
+    }
   })
+
+  store.setPrimary('ethereum', 137, { connected: false })
+  store.activateNetwork('ethereum', 137, true)
```

Comment:
is this better than the mocking?  Certainly takes up more space, though I guess you get more familiar with the APIs and don't have to reference the underlying data structure.  Looking at the codebase there's a mishmash of both approaches.

Reply:
the main reason I did it is so that we can modify the state in each individual test to meet the needs of that test, previously the state was hardcoded and so I started trying to update it so that we could update the mocked state and landed on this approach. I think also in general when you can do something without mocking without sacrificing speed or readability I like to default to that path